src/lib/io/sol-i2c-common.c

```diff
@@ -23,20 +23,259 @@
+    struct sol_vector *set;
+    void (*cb)(void *cb_data, ssize_t status);
+    void *cb_data;
+    uint32_t delay;
+    uint8_t addr;
+    bool delete_me;
+};
+
+static bool i2c_dispatcher_exec_op(void *data);
+
+static void
+i2c_dispatcher_end_set(struct sol_i2c_dispatcher *dispatcher, ssize_t end_status)
+{
+    struct sol_i2c_op_set *set = sol_ptr_vector_get(&dispatcher->queue, 0);
+
+    if (!set->delete_me && set->cb)
+        set->cb(set->cb_data, end_status);
+
+    free(set);
+    sol_ptr_vector_del(&dispatcher->queue, 0);
+
+    dispatcher->set_idx = 0;
+}
+
+static void
+i2c_dispatcher_op_done(void *cb_data, struct sol_i2c *i2c, uint8_t reg, uint8_t *data,
+    ssize_t status)
+{
+    struct sol_i2c_shared *shared = cb_data;
+    struct sol_i2c_dispatcher *dispatcher = &shared->dispatcher;
+    struct sol_i2c_op_set *op_set = sol_ptr_vector_get(&dispatcher->queue, 0);
+
+    dispatcher->pending = NULL;
+    dispatcher->retry = 0;
+
+    if (!op_set->delete_me) {
+        if (status <= 0) {
+            SOL_ERR("[bus=%d addr=0x%02x reg=0x%02x] I2C operation failed!",
+                sol_i2c_bus_get(i2c), op_set->addr, reg);
+        } else if (++dispatcher->set_idx < op_set->set->len) {
+            goto proceed;
+        }
+    }
+
+    i2c_dispatcher_end_set(dispatcher, status);
+    op_set = sol_ptr_vector_get(&dispatcher->queue, 0);
+
+    if (!op_set)
+        return; // We finished for now
+
+proceed:
+    if (dispatcher->timer)
+        SOL_ERR("Dispatcher timer should be NULL in this point.");
+    dispatcher->timer = sol_timeout_add(op_set->delay, i2c_dispatcher_exec_op, shared);
+    if (!dispatcher->timer) {
+        SOL_ERR("Failed to schedule I2C operation.");
+        status = -1;
+        if (op_set->cb)
+            op_set->cb(op_set->cb_data, status);
+    }
+}
+
+static bool
+i2c_dispatcher_exec_op(void *data)
+{
+    struct sol_i2c_op *op;
+    struct sol_i2c_shared *shared = data;
+    struct sol_i2c_dispatcher *dispatcher = &shared->dispatcher;
+    struct sol_i2c_op_set *op_set = sol_ptr_vector_get(&dispatcher->queue, 0);
+
+    if (!op_set)
+        goto exit;
+
+    if (sol_i2c_busy(shared->i2c) || dispatcher->pending) {
+        if (++dispatcher->retry >= SOL_I2C_MAX_RETRIES) {
+            SOL_ERR("Failed to schedule I2C operation.");
+            goto exit;
+        }
+
+        return true;
+    }
+
+    if (!sol_i2c_set_slave_address(shared->i2c, op_set->addr)) {
+        SOL_ERR("Failed to set slave address 0x%02x on I2C bus: %d.", op_set->addr,
+            sol_i2c_bus_get(shared->i2c));
+        i2c_dispatcher_end_set(dispatcher, -1);
+        return true;
+    }
+
+    op = sol_vector_get(op_set->set, dispatcher->set_idx);
+
+    dispatcher->pending = (op->type == SOL_I2C_WRITE) ?
+        sol_i2c_write_register(shared->i2c, op->reg, &op->value, 1,
+        i2c_dispatcher_op_done, shared) :
+        sol_i2c_read_register(shared->i2c, op->reg, &op->value, 1,
+        i2c_dispatcher_op_done, shared);
```

Comment:
You should state that this API is solely for 1 byte read/writes or, maybe, extend it to support more bytes as well